### ISSUE TITLE
Copy button broken on WebGL?

### ISSUE THREAD

**Issue by dbirman** [VirtualBrainLab/Pinpoint#245]

**Comment by dbirman** [comment]
WebGL disables buffer access https://answers.unity.com/questions/1364513/how-to-make-copypast-work-on-webgl.html

**Comment by dbirman** [comment]
Alternative: create a json serializer/de-serializer for ProbeManager and return / load this?

**Comment by dbirman** [comment]
Attempting fix in 0.8.4 using browser work-around: 84ff53d56295954c772d16f7ed811483f45812c6

**Comment by dbirman** [comment]
Blocked due to lack of click enabling action: `Uncaught (in promise) DOMException: Clipboard write was blocked due to lack of user activation.`

Note sure how to fix this...

**Comment by dbirman** [comment]
Alternate solution from https://forum.unity.com/threads/how-do-i-let-the-user-load-an-image-from-their-harddrive-into-a-webgl-app.380985/

2) Uploading an image using UI button

This one is quite tricky. The idea would be of course to make the html browse button hidden, and simulate its click programmatically. We are however facing a security restriction here, specifically, it is not possible to launch the <input type="file"> onclick event handler programmatically, unless you make the click() call from another event handler initiated by the user. And as in most complex systems, in Unity WebGL input events are not processed directly but instead are passed through an intermediate queue. This means that the initial JavaScript event handler initiated by the user has been already processed by the time when you receive the OnClick event in the managed code.
Nevertheless, there is a trick that we can perform to make it work: as soon as the user pushes the UI button down, we can register onclick JavaScript event for the whole WebGL canvas, so that when the user releases the button, we will get a user-initiated onclick JavaScript event, and perform the input click simulation in the handler.

You will need the following Assets/Plugins/ImageUploader.jslib plugin:
Code (JavaScript):
```
    var ImageUploaderPlugin = {
      ImageUploaderCaptureClick: function() {
        if (!document.getElementById('ImageUploaderInput')) {
          var fileInput = document.createElement('input');
          fileInput.setAttribute('type', 'file');
          fileInput.setAttribute('id', 'ImageUploaderInput');
          fileInput.style.visibility = 'hidden';
          fileInput.onclick = function (event) {
            this.value = null;
          };
          fileInput.onchange = function (event) {
            SendMessage('Canvas', 'FileSelected', URL.createObjectURL(event.target.files[0]));
          }
          document.body.appendChild(fileInput);
        }
        var OpenFileDialog = function() {
          document.getElementById('ImageUploaderInput').click();
          document.getElementById('canvas').removeEventListener('click', OpenFileDialog);
        };
        document.getElementById('canvas').addEventListener('click', OpenFileDialog, false);
      }
    };
    mergeInto(LibraryManager.library, ImageUploaderPlugin);
```
And the following script to interact with it:
Code (CSharp):
```
    using UnityEngine;
    using System.Collections;
    using System.Runtime.InteropServices;
     
    public class CanvasScript : MonoBehaviour {
        [DllImport("__Internal")]
        private static extern void ImageUploaderCaptureClick();
     
        IEnumerator LoadTexture (string url) {
            WWW image = [new](http://www.google.com/search?q=new+msdn.microsoft.com) WWW (url);
            yield return image;
            Texture2D texture = [new](http://www.google.com/search?q=new+msdn.microsoft.com) Texture2D (1, 1);
            image.LoadImageIntoTexture (texture);
            Debug.Log ("Loaded image size: " + texture.width + "x" + texture.height);
        }
     
        void FileSelected (string url) {
            StartCoroutine(LoadTexture (url));
        }
     
        public void OnButtonPointerDown () {
            ImageUploaderCaptureClick ();
        }
    }
```
The OnButtonPointerDown should be triggered by the Pointer Down event (available through the Event Trigger component, that you can attach to your UI button)

The second method is of course not 100% reliable (for example, user can push the button, move the mouse outside the canvas, then release the button, so that the attached onclick event handler remains unprocessed), but for most use cases it should be quite appropriate.

**Comment by dbirman** [comment]
Testing in develop release ff12b2a9a09214ede79822e696e0fe773f723d05

**Comment by dbirman** [comment]
Not working because:

- [ ] UI button in Unity is triggered from the click function, needs to be triggered by OnMouseDown
- [ ] Somehow the string is getting mangled as it gets passed from C# to Javascript?

**Comment by dbirman** [comment]
New attempt to fix 52e438ea545996c92fbb0989a2983f65a4b71862

**Comment by dbirman** [comment]
Button works now, but text string is wrong

**Comment by dbirman** [comment]
Fixed with 8801bc262344bdaa64202ebc37586076d088fcd3
